src/Pos/Client/Txp/Util.hs

```diff
@@ -174,27 +221,35 @@ makeRedemptionTx rsk txInputs = makeAbstractTx mkWit (map ((), ) txInputs)
             }
 
 type FlatUtxo = [(TxIn, TxOutAux)]
-type InputPicker = StateT (Coin, FlatUtxo) (Either Text)
+type InputPicker = StateT (Coin, FlatUtxo) (Either TxError)
 
--- | Given Utxo, desired source addresses and desired outputs, prepare lists
--- of correct inputs and outputs to form a transaction
-prepareInpsOuts ::
-       MonadError Text m
+-- | Given filtered Utxo, desired outputs and fee size,
+-- prepare correct inputs and outputs for transaction
+-- (and tell how much to send to remaining address)
+prepareTxRaw
+    :: MonadError TxError m
     => Utxo
-    -> NonEmpty Address
     -> TxOutputs
-    -> m (TxOwnedInputs Address, TxOutputs)
-prepareInpsOuts utxo addrs outputs = do
+    -> TxFee
+    -> m TxRaw
+prepareTxRaw utxo outputs (TxFee fee) = do
     when (totalMoney == mkCoin 0) $
-        throwError "Attempted to send 0 money"
+        throwTxError "Attempted to send 0 money"
     futxo <- either throwError pure $
-        evalStateT (pickInputs []) (totalMoney, sortedUnspent)
+        evalStateT (pickInputs []) (totalMoneyWithFee, sortedUnspent)
     case nonEmpty futxo of
-        Nothing       -> throwError "Failed to prepare inputs!"
-        Just inputsNE -> pure (map formTxInputs inputsNE, outputs)
+        Nothing       -> throwTxError "Failed to prepare inputs!"
+        Just inputsNE -> do
+            let totalTxAmount = unsafeSumTxOuts $ map snd inputsNE
+                trInputs = map formTxInputs inputsNE
+                trOutputs = outputs
+                trRemaining = totalTxAmount `unsafeSubCoin` totalMoneyWithFee
+            pure TxRaw {..}
   where
-    totalMoney = unsafeIntegerToCoin $ sumCoins $ map (txOutValue . toaOut) outputs
-    allUnspent = M.toList $ filterUtxoByAddrs (toList addrs) utxo
+    unsafeSumTxOuts = unsafeIntegerToCoin . sumCoins . map (txOutValue . toaOut)
+    totalMoney = unsafeSumTxOuts outputs
+    totalMoneyWithFee = totalMoney `unsafeAddCoin` fee
+    allUnspent = M.toList utxo
```

Comment:
and it's subjective, but IMHO a function w/ several named parameters looks better than tuple-state and access to them via `_1`, `_2`... 
because non-familar reader is forced to see what this tuple means every time when he sees `_1`, `_2`, `_3`